sparta/sparta/functional/RegisterBits.hpp

```diff
@@ -0,0 +1,704 @@
+// <RegisterBits> -*- C++ -*-
+#pragma once
+
+#include <cinttypes>
+#include <array>
+#include <functional>
+#include <limits>
+#include <string.h>
+#include <algorithm>
+
+#include "sparta/utils/SpartaAssert.hpp"
+
+namespace sparta
+{
+    /**
+     * \class RegisterBits
+     *
+     * This class is used in conjuntion with sparta::RegisterBase to
+     * quickly write masked registers of sizes between 1 and 512
+     * bytes.  This class replaces the use of BitArray.
+     *
+     * The class works by assuming register data is handed to it via a
+     * char array.  The class will "view" into this data until it's
+     * requested to modify it.  When it's modified, the data will be
+     * copied to local storage within this class.  The user of the
+     * class is responsible for writing the data back to the original
+     * storage.
+     */
+    class RegisterBits
+    {
+        template<typename SizeT, typename Op>
+        static RegisterBits bitMerge_(const RegisterBits & in, const RegisterBits & rh_bits) {
+            RegisterBits final_value(in.num_bytes_);
+            *reinterpret_cast<SizeT*>(final_value.local_data_) =
+                Op()(*reinterpret_cast<const SizeT*>(in.remote_data_),
+                     *reinterpret_cast<const SizeT*>(rh_bits.remote_data_));
+            return final_value;
+        }
+
+        template<typename SizeT>
+        static RegisterBits bitNot_(const RegisterBits & in) {
+            RegisterBits final_value(in.num_bytes_);
+            *reinterpret_cast<SizeT*>(final_value.local_data_) =
+                std::bit_not<SizeT>()(*reinterpret_cast<const SizeT*>(in.remote_data_));
+            return final_value;
+        }
+
+        template<typename SizeT>
+        static RegisterBits bitShiftRight_(const RegisterBits & in, uint32_t amount) {
+            RegisterBits final_value(in.num_bytes_);
+            *reinterpret_cast<SizeT*>(final_value.local_data_) =
+                (*reinterpret_cast<const SizeT*>(in.remote_data_)) >> amount;
+            return final_value;
+        }
+
+        template<typename SizeT>
+        static RegisterBits bitShiftLeft_(const RegisterBits & in, uint32_t amount) {
+            RegisterBits final_value(in.num_bytes_);
+            *reinterpret_cast<SizeT*>(final_value.local_data_) =
+                (*reinterpret_cast<const SizeT*>(in.remote_data_)) << amount;
+            return final_value;
+        }
+
+        // Copy the remote register data locally.
+        void convert_() {
+            if(nullptr == local_data_) {
+                local_data_ = local_storage_.data();
+                ::memset(local_data_, 0, local_storage_.size());
+                ::memcpy(local_data_, remote_data_, num_bytes_);
+                remote_data_ = local_data_;
```

Comment:
Believe it or not, doing this `memset` here is _faster_ than doing `local_data_ = {0}` initialization at the member declaration.